File: deploy/kind/probes/api_client_probe.py
# SPDX-License-Identifier: Apache-2.0
"""Kind API/client compatibility surface probe.

This is the containerized counterpart to `just api-client-local-smoke`: it runs
inside the maintained Kind network and proves the protocol/catalog surfaces that
Python/API/BI clients need before each named client gets its own heavier probe.
"""

from __future__ import annotations

import sys

import psycopg2

from probe_common import (
    ORIGIN_WKB,
    ONE_WKB,
    pg_connect,
    quote_ident,
    require,
    require_equal,
    table_name,
)


def point_wkb(x: float, y: float) -> bytes:
    import struct

    return b"\x01" + struct.pack("<I", 1) + struct.pack("<dd", x, y)


def seed(conn, table: str) -> None:
    table_ref = f"public.{quote_ident(table)}"
    with conn.cursor() as cur:
        cur.execute(
            f"CREATE TABLE {table_ref} "
            "(id INT, geom BINARY, name TEXT, category TEXT)"
        )
        cur.execute(
            f"INSERT INTO {table_ref} (id, geom, name, category) VALUES "
            f"(1, X'{ORIGIN_WKB}', 'origin', 'alpha'), "
            f"(2, X'{ONE_WKB}', 'one', 'alpha'), "
            "(3, X'010100000000000000000014400000000000001440', 'far', 'beta')"
        )


def assert_psycopg_surface(conn, table: str) -> None:
    with conn.cursor() as cur:
        cur.execute("SELECT ST_AsText(ST_GeomFromText(%s))", ("POINT(3 4)",))
        require_equal(cur.fetchone()[0], "POINT(3 4)", "text WKT parameter")

        cur.execute(
            "SELECT ST_AsText(ST_GeomFromWKB(%s))",
            (psycopg2.Binary(point_wkb(3.0, 4.0)),),
        )
        require_equal(cur.fetchone()[0], "POINT(3 4)", "binary WKB parameter")

        cur.execute(
            f"SELECT ST_AsEWKB(geom) FROM public.{quote_ident(table)} WHERE id = 1"
        )
        ewkb = cur.fetchone()[0]
        require(bytes(ewkb) == bytes.fromhex(ORIGIN_WKB), "ST_AsEWKB bytes changed")
    print("api_client_psycopg_surface text_param=True binary_wkb=True ewkb=True")


def assert_reflection_surface(conn, table: str) -> int:
    with conn.cursor() as cur:
        cur.execute(
            "SELECT COUNT(*) FROM information_schema.tables "
            "WHERE table_schema = 'public' AND table_name = %s",
            (table,),
        )
        require_equal(cur.fetchone()[0], 1, "information_schema table count")
        cur.execute(
            "SELECT column_name FROM information_schema.columns "
            "WHERE table_schema = 'public' AND table_name = %s "
            "ORDER BY ordinal_position",
            (table,),
        )
        columns = [row[0] for row in cur.fetchall()]
    for expected in ("id", "geom", "name", "category"):
        require(expected in columns, f"missing reflected column {expected}: {columns!r}")
    require(
        all(not column.startswith("_qg_") for column in columns),
        f"hidden layout columns leaked through public reflection: {columns!r}",
    )
    print(f"api_client_sqlalchemy_surface columns={columns!r}")
    return len(columns)


def assert_geopandas_surface(conn, table: str) -> int:
    with conn.cursor() as cur:
        cur.execute(
            f"SELECT id, name, ST_AsEWKB(geom) AS geom "
            f"FROM public.{quote_ident(table)} ORDER BY id"
        )
        rows = cur.fetchall()
    require_equal(len(rows), 3, "GeoPandas-style feature count")
    require(bytes(rows[0][2]) == bytes.fromhex(ORIGIN_WKB), "first WKB changed")
    print("api_client_geopandas_surface feature_count=3 wkb=True crs_documented=srid0")
    return len(rows)


def assert_pgfeatureserv_surface(conn, table: str) -> int:
    with conn.cursor() as cur:
        cur.execute(
            f"SELECT id, name FROM public.{quote_ident(table)} "
            "WHERE ST_Intersects("
            "ST_GeomFromWKB(geom), "
            "ST_GeomFromWKB(ST_MakeEnvelope(-0.5, -0.5, 0.5, 0.5, 3857))"
            ") ORDER BY id"
        )
        rows = cur.fetchall()
    require_equal(rows, [(1, "origin")], "pg_featureserv bbox rows")
    print("api_client_pgfeatureserv_surface bbox_count=1 properties=True")
    return len(rows)


def assert_bi_surface(conn, table: str) -> int:
    with conn.cursor() as cur:
        cur.execute(
            f"SELECT category, COUNT(*) FROM public.{quote_ident(table)} "
            "GROUP BY category ORDER BY category"
        )
        rows = cur.fetchall()
    require_equal(rows, [("alpha", 2), ("beta", 1)], "BI grouped aggregate")
    print(f"api_client_bi_surface groups={len(rows)} grouped={rows!r}")
    return len(rows)


def assert_mvt_surface(conn, table: str) -> int:
    with conn.cursor() as cur:
        cur.execute(
            "SELECT ST_AsMVT("
            "ST_AsMVTGeom(geom, ST_MakeEnvelope(-1.0, -1.0, 2.0, 2.0, 3857), 4096, 64, true)"
            f") FROM public.{quote_ident(table)} WHERE id <= 2"
        )
        tile = cur.fetchone()[0]
    tile_bytes = len(bytes(tile))
    require(tile_bytes > 0, "MVT tile was empty")
    print(f"api_client_mvt_surface tile_bytes={tile_bytes}")
    return tile_bytes


def main() -> int:
    table = table_name("api_client_probe")
    conn = pg_connect()
    conn.autocommit = True
    try:
        seed(conn, table)
        assert_psycopg_surface(conn, table)
        reflected_columns = assert_reflection_surface(conn, table)
        feature_count = assert_geopandas_surface(conn, table)
        bbox_count = assert_pgfeatureserv_surface(conn, table)
        groups = assert_bi_surface(conn, table)
        tile_bytes = assert_mvt_surface(conn, table)
    finally:
        conn.close()

    print(
        "api_client_summary "
        f"feature_count={feature_count} "
        f"reflected_columns={reflected_columns} "
        f"bbox_count={bbox_count} "
        f"groups={groups} "
        f"tile_bytes={tile_bytes}"
    )
    print("api_client_probe_ok True")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/quackgis-server/examples/api_client_probe.rs
// SPDX-License-Identifier: Apache-2.0
//! Local API/client surface probe for Python/API/BI-style compatibility work.
//!
//! This intentionally uses `tokio-postgres` rather than heavy client runtimes. It
//! captures the PostgreSQL wire/catalog/SQL surfaces that psycopg, SQLAlchemy,
//! GeoPandas, pg_featureserv-style readers, BI tools, and MVT consumers need
//! before a containerized client probe is worth maintaining.

use anyhow::{Context, Result, ensure};
use tokio_postgres::{NoTls, types::Type};

#[derive(Debug)]
struct Config {
    host: String,
    port: u16,
}

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let config = Config::from_env_and_args()?;
    let conn_str = format!(
        "host={} port={} user=postgres dbname=quackgis",
        config.host, config.port
    );
    let (client, connection) = tokio_postgres::connect(&conn_str, NoTls)
        .await
        .with_context(|| format!("connect to QuackGIS at {}:{}", config.host, config.port))?;
    tokio::spawn(async move {
        if let Err(err) = connection.await {
            eprintln!("postgres connection task failed: {err}");
        }
    });

    seed_probe_table(&client).await?;
    assert_psycopg_surface(&client).await?;
    assert_sqlalchemy_surface(&client).await?;
    assert_geopandas_surface(&client).await?;
    assert_pgfeatureserv_surface(&client).await?;
    assert_bi_surface(&client).await?;
    assert_mvt_surface(&client).await?;

    println!("api_client_probe_ok True");
    Ok(())
}

impl Config {
    fn from_env_and_args() -> Result<Self> {
        let mut host = std::env::var("QUACKGIS_HOST").unwrap_or_else(|_| "127.0.0.1".to_string());
        let mut port = std::env::var("QUACKGIS_PORT")
            .ok()
            .and_then(|value| value.parse::<u16>().ok())
            .unwrap_or(5434);
        let mut args = std::env::args().skip(1);
        while let Some(arg) = args.next() {
            match arg.as_str() {
                "--host" => host = args.next().context("--host requires a value")?,
                "--port" => {
                    port = args
                        .next()
                        .context("--port requires a value")?
                        .parse()
                        .context("--port must be a u16")?;
                }
                "--help" | "-h" => {
                    println!(
                        "Usage: cargo run -p quackgis-server --example api_client_probe -- [--host HOST] [--port PORT]"
                    );
                    std::process::exit(0);
                }
                other => anyhow::bail!("unknown argument: {other}"),
            }
        }
        Ok(Self { host, port })
    }
}

async fn seed_probe_table(client: &tokio_postgres::Client) -> Result<()> {
    client
        .batch_execute(
            "CREATE TABLE public.api_probe_points (
                 id INT,
                 geom BINARY,
                 name TEXT,
                 category TEXT
             );
             INSERT INTO public.api_probe_points (id, geom, name, category) VALUES
                 (1, X'010100000000000000000000000000000000000000', 'origin', 'alpha'),
                 (2, X'0101000000000000000000F03F000000000000F03F', 'one', 'alpha'),
                 (3, X'010100000000000000000014400000000000001440', 'far', 'beta');",
        )
        .await
        .context("seed api_probe_points")?;
    Ok(())
}

async fn assert_psycopg_surface(client: &tokio_postgres::Client) -> Result<()> {
    let text_param: String = client
        .query_one("SELECT ST_AsText(ST_GeomFromText($1))", &[&"POINT(3 4)"])
        .await
        .context("text parameter WKT roundtrip")?
        .get(0);
    ensure!(text_param == "POINT(3 4)", "text param got {text_param}");

    let binary_stmt = client
        .prepare_typed("SELECT ST_AsText(ST_GeomFromWKB($1))", &[Type::BYTEA])
        .await
        .context("prepare typed bytea parameter")?;
    let binary_param: String = client
        .query_one(&binary_stmt, &[&point_wkb(3.0, 4.0)])
        .await
        .context("binary WKB parameter roundtrip")?
        .get(0);
    ensure!(
        binary_param == "POINT(3 4)",
        "binary param got {binary_param}"
    );

    let ewkb: Vec<u8> = client
        .query_one(
            "SELECT ST_AsEWKB(geom) FROM public.api_probe_points WHERE id = 1",
            &[],
        )
        .await
        .context("ST_AsEWKB readback")?
        .get(0);
    ensure!(ewkb == point_wkb(0.0, 0.0), "EWKB/WKB bytes changed");
    println!("api_client_psycopg_surface text_param=True binary_wkb=True ewkb=True");
    Ok(())
}

async fn assert_sqlalchemy_surface(client: &tokio_postgres::Client) -> Result<()> {
    let table_count: i64 = client
        .query_one(
            "SELECT COUNT(*) FROM information_schema.tables
             WHERE table_schema = 'public' AND table_name = 'api_probe_points'",
            &[],
        )
        .await
        .context("information_schema table reflection")?
        .get(0);
    ensure!(table_count == 1, "table reflection count={table_count}");

    let columns = client
        .query(
            "SELECT column_name FROM information_schema.columns
             WHERE table_schema = 'public' AND table_name = 'api_probe_points'
             ORDER BY ordinal_position",
            &[],
        )
        .await
        .context("information_schema column reflection")?
        .into_iter()
        .map(|row| row.get::<_, String>(0))
        .collect::<Vec<_>>();
    for expected in ["id", "geom", "name", "category"] {
        ensure!(
            columns.iter().any(|column| column == expected),
            "reflected columns {columns:?} missing {expected}"
        );
    }
    ensure!(
        columns.iter().all(|column| !column.starts_with("_qg_")),
        "reflected public columns leaked hidden layout columns: {columns:?}"
    );
    println!("api_client_sqlalchemy_surface table=True columns={columns:?}");
    Ok(())
}

async fn assert_geopandas_surface(client: &tokio_postgres::Client) -> Result<()> {
    let rows = client
        .query(
            "SELECT id, name, ST_AsEWKB(geom) AS geom
             FROM public.api_probe_points
             ORDER BY id",
            &[],
        )
        .await
        .context("GeoPandas-style WKB row query")?;
    ensure!(rows.len() == 3, "GeoPandas row count={}", rows.len());
    let first_geom: Vec<u8> = rows[0].get(2);
    ensure!(first_geom == point_wkb(0.0, 0.0), "first WKB changed");
    println!("api_client_geopandas_surface feature_count=3 wkb=True crs_documented=srid0");
    Ok(())
}

async fn assert_pgfeatureserv_surface(client: &tokio_postgres::Client) -> Result<()> {
    let rows = client
        .query(
            "SELECT id, name
             FROM public.api_probe_points
             WHERE ST_Intersects(
                 ST_GeomFromWKB(geom),
                 ST_GeomFromWKB(ST_MakeEnvelope(-0.5, -0.5, 0.5, 0.5, 3857))
             )
             ORDER BY id",
            &[],
        )
        .await
        .context("pg_featureserv-style bbox query")?;
    let ids = rows
        .into_iter()
        .map(|row| row.get::<_, i32>(0))
        .collect::<Vec<_>>();
    ensure!(ids == vec![1], "bbox ids={ids:?}");
    println!("api_client_pgfeatureserv_surface bbox_count=1 properties=True");
    Ok(())
}

async fn assert_bi_surface(client: &tokio_postgres::Client) -> Result<()> {
    let grouped = client
        .query(
            "SELECT category, COUNT(*) AS n
             FROM public.api_probe_points
             GROUP BY category
             ORDER BY category",
            &[],
        )
        .await
        .context("BI grouped aggregate query")?
        .into_iter()
        .map(|row| (row.get::<_, String>(0), row.get::<_, i64>(1)))
        .collect::<Vec<_>>();
    ensure!(
        grouped == vec![("alpha".to_string(), 2), ("beta".to_string(), 1)],
        "grouped aggregate got {grouped:?}"
    );
    println!("api_client_bi_surface grouped={grouped:?}");
    Ok(())
}

async fn assert_mvt_surface(client: &tokio_postgres::Client) -> Result<()> {
    let tile: Vec<u8> = client
        .query_one(
            "SELECT ST_AsMVT(
                  ST_AsMVTGeom(
                   geom,
                   ST_MakeEnvelope(-1.0, -1.0, 2.0, 2.0, 3857),
                   4096,
                   64,
                    true
                  ),
                  'api_probe_points',
                  4096,
                  name,
                  category
               )
               FROM public.api_probe_points
               WHERE id <= 2",
            &[],
        )
        .await
        .context("MVT tile query")?
        .get(0);
    ensure!(!tile.is_empty(), "MVT tile was empty");
    for expected in ["api_probe_points", "name", "origin", "category", "alpha"] {
        ensure!(
            contains_bytes(&tile, expected.as_bytes()),
            "MVT tile missing expected attribute/layer token {expected:?}"
        );
    }
    println!(
        "api_client_mvt_surface tile_bytes={} attributes=True",
        tile.len()
    );
    Ok(())
}

fn contains_bytes(haystack: &[u8], needle: &[u8]) -> bool {
    haystack
        .windows(needle.len())
        .any(|window| window == needle)
}

fn point_wkb(x: f64, y: f64) -> Vec<u8> {
    let mut out = Vec::with_capacity(21);
    out.push(1);
    out.extend_from_slice(&1_u32.to_le_bytes());
    out.extend_from_slice(&x.to_le_bytes());
    out.extend_from_slice(&y.to_le_bytes());
    out
}
